airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/abstract_stream.py

```diff
@@ -0,0 +1,152 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+import logging
+from abc import ABC, abstractmethod
+from functools import lru_cache
+from typing import Any, Iterable, List, Mapping, Optional, Tuple, Union
+
+from airbyte_cdk.models import SyncMode
+from airbyte_cdk.sources.streams import Stream
+from airbyte_cdk.sources.streams.core import StreamData
+from airbyte_cdk.sources.utils import casing
+from airbyte_cdk.sources.utils.slice_logger import SliceLogger
+from deprecated.classic import deprecated
+
+
+@deprecated("This class is experimental. Use at your own risk.")
+class AbstractStream(ABC):
+    """
+    AbstractStream is an experimental interface for streams developed as part of the Concurrent CDK.
+    This interface is not yet stable and may change in the future. Use at your own risk.
+
+    Why create a new interface instead of adding concurrency capabilities the existing Stream?
+    We learnt a lot since the initial design of the Stream interface, and we wanted to take the opportunity to improve.
+
+    High level, the changes we are targeting are:
+    - Removing superfluous or leaky parameters from the methods' interfaces
+    - Using composition instead of inheritance to add new capabilities
+
+    To allow us to iterate fast while ensuring backwards compatibility, we are creating a new interface with a facade object that will bridge the old and the new interfaces.
+    Source connectors that which to leverage concurrency need to implement this new interface. An example will be available shortly
+
+    Current restrictions on sources that implement this interface. Not all of these restrictions will be lifted in the future, but most will as we iterate on the design.
+    - Only full refresh is supported. This will be addressed in the future.
+    - The read method does not accept a cursor_field. Streams must be internally aware of the cursor field to use. User-defined cursor fields can be implemented by modifying the connector's main method to instantiate the streams with the configured cursor field.
+    - Streams cannot return user-friendly messages by overriding Stream.get_error_display_message. This will be addressed in the future.
```

Comment:
this will require a follow up issue

Reply:
@girarda I'm not sure I understand how instantiating the stream with the configured cursor field is different from passing it to `read_full_records`?
I think I understand what you tried to do here, to make stream a stateful object, but the current interface, i.e. existence of method `read_full_records` kind of assume that we can re-use the same stream instance in different modes (which don't think is a good idea)

Reply:
Yes, it is already possible and should be discouraged, IMO, because some streams might have a state outside of reading functions (on the class/instance level) nowadays. The Abstract class is much more straightforward.